js/modules/loki_message_api.js

```diff
@@ -32,109 +34,143 @@ class LokiServer {
+      if (
+        options.responseType === 'json' &&
+        response.headers.get('Content-Type') === 'application/json'
+      ) {
+        result = await response.json();
+      } else if (options.responseType === 'arraybuffer') {
+        result = await response.buffer();
+      } else {
+        result = await response.text();
+      }
 
-    if (response.status >= 0 && response.status < 400) {
+      if (response.status >= 0 && response.status < 400) {
+        return result;
+      }
+      log.error(options.type, options.url, response.status, 'Error sending message');
+      throw HTTPError('sendMessage: error response', response.status, result);
+    });
+    try {
+      // TODO: Possibly change this to require more than a single response?
+      const result = await firstOf(requests);
       return result;
+    } catch(err) {
+      throw err;
     }
-    log.error(options.type, options.url, response.status, 'Error');
-    throw HTTPError('sendMessage: error response', response.status, result);
   }
 
-  async retrieveMessages(pubKey) {
-    // Hardcoded to use a single node/server for now
-    const currentNode = this.nodes[0];
-
-    const options = {
-      url: `${currentNode.url}/retrieve`,
-      type: 'GET',
-      responseType: 'json',
-      timeout: undefined,
-    };
-
-    const headers = {
-      'X-Loki-recipient': pubKey,
-    };
-
-    if (currentNode.lastHash) {
-      headers['X-Loki-last-hash'] = currentNode.lastHash;
-    }
-
-    const fetchOptions = {
-      method: options.type,
-      headers,
-      timeout: options.timeout,
-    };
+  async retrieveMessages(callback) {
+    let ourSwarmNodes = await window.LokiSnodeAPI.getOurSwarmNodes();
+    const ourKey = window.textsecure.storage.user.getNumber();
+    let completedRequests = 0;
+
+    const doRequest = async (nodeUrl, nodeData) => {
+      const options = {
+        url: `${nodeUrl}${this.messageServerPort}/retrieve`,
+        type: 'GET',
+        responseType: 'json',
+        timeout: undefined,
+      };
+
+      const headers = {
+        'X-Loki-recipient': ourKey,
+      };
+
+      if (nodeData.lastHash) {
+        headers['X-Loki-last-hash'] = nodeData.lastHash;
+      }
 
-    let response;
-    try {
-      response = await fetch(options.url, fetchOptions);
-    } catch (e) {
-      log.error(options.type, options.url, 0, 'Error');
-      throw HTTPError('fetch error', 0, e.toString());
-    }
+      const fetchOptions = {
+        method: options.type,
+        headers,
+        timeout: options.timeout,
+      };
+      let response;
+      try {
+        response = await fetch(options.url, fetchOptions);
+      } catch (e) {
+        // TODO: Maybe we shouldn't immediately delete?
+        log.error(options.type, options.url, 0, `Error retrieving messages from ${nodeUrl}`);
+        window.LokiSnodeAPI.unreachableNode(ourKey, nodeUrl);
+        throw HTTPError('fetch error', 0, e.toString());
+      }
+      completedRequests += 1;
```

Comment:
Do we want this here or below after decoding the result?

Reply:
I think at this point we have confirmed that the node has responded (is online) right? Tho if the decoding can throw errors (not 100% about this?) maybe we want to label the node as unreachable anyway?

I'll move it